File: qeli-shared/QeliShared/Geo/GeoSiteIndex.cs
namespace Qeli.Shared.Geo;

/// <summary>v2ray Domain.Type</summary>
public enum GeoDomainType : byte
{
    Plain = 0,
    Regex = 1,
    RootDomain = 2,
    Full = 3,
}

public readonly record struct GeoDomainRule(GeoDomainType Type, string Value);

/// <summary>Loaded subset of geosite.dat for selected tags (case-insensitive).</summary>
public sealed class GeoSiteIndex
{
    private readonly Dictionary<string, List<GeoDomainRule>> _byTag =
        new(StringComparer.OrdinalIgnoreCase);

    public IReadOnlyCollection<string> Tags => _byTag.Keys;

    public static GeoSiteIndex Load(string path, IEnumerable<string> tags)
    {
        var want = new HashSet<string>(tags, StringComparer.OrdinalIgnoreCase);
        var bytes = File.ReadAllBytes(path);
        var index = new GeoSiteIndex();
        var reader = new ProtoReader(bytes);
        while (reader.HasMore)
        {
            if (!reader.TryReadTag(out int field, out int wt)) break;
            if (field != 1 || wt != 2) { reader.Skip(wt); continue; }
            ParseGeoSite(reader.ReadBytes(), want, index);
        }
        return index;
    }

    private static void ParseGeoSite(ReadOnlySpan<byte> msg, HashSet<string> want, GeoSiteIndex index)
    {
        string? code = null;
        var domains = new List<GeoDomainRule>();
        var r = new ProtoReader(msg);
        while (r.HasMore)
        {
            if (!r.TryReadTag(out int field, out int wt)) break;
            switch (field)
            {
                case 1 when wt == 2:
                    code = r.ReadString();
                    break;
                case 2 when wt == 2:
                    domains.Add(ParseDomain(r.ReadBytes()));
                    break;
                default:
                    r.Skip(wt);
                    break;
            }
        }
        if (code == null || !want.Contains(code)) return;
        if (!index._byTag.TryGetValue(code, out var list))
        {
            list = new List<GeoDomainRule>();
            index._byTag[code] = list;
        }
        list.AddRange(domains);
    }

    private static GeoDomainRule ParseDomain(ReadOnlySpan<byte> msg)
    {
        GeoDomainType type = GeoDomainType.Plain;
        string value = "";
        var r = new ProtoReader(msg);
        while (r.HasMore)
        {
            if (!r.TryReadTag(out int field, out int wt)) break;
            switch (field)
            {
                case 1 when wt == 0:
                    type = (GeoDomainType)(int)r.ReadVarint();
                    break;
                case 2 when wt == 2:
                    value = r.ReadString();
                    break;
                default:
                    r.Skip(wt);
                    break;
            }
        }
        return new GeoDomainRule(type, value);
    }

    public bool MatchAny(string host, params string[] tags)
    {
        if (string.IsNullOrWhiteSpace(host)) return false;
        host = host.Trim().TrimEnd('.').ToLowerInvariant();
        foreach (var tag in tags)
        {
            if (!_byTag.TryGetValue(tag, out var rules)) continue;
            foreach (var rule in rules)
            {
                if (MatchOne(host, rule)) return true;
            }
        }
        return false;
    }

    private static bool MatchOne(string host, GeoDomainRule rule)
    {
        var v = rule.Value.Trim().ToLowerInvariant();
        if (v.Length == 0) return false;
        return rule.Type switch
        {
            GeoDomainType.Full => host == v,
            GeoDomainType.RootDomain => host == v || host.EndsWith("." + v, StringComparison.Ordinal),
            GeoDomainType.Plain => host.Contains(v, StringComparison.Ordinal),
            GeoDomainType.Regex => SafeRegex(host, v),
            _ => false,
        };
    }

    private static bool SafeRegex(string host, string pattern)
    {
        try { return System.Text.RegularExpressions.Regex.IsMatch(host, pattern); }
        catch { return false; }
    }
}

File: qeli-android/app/src/main/kotlin/com/qeli/geo/GeoSiteIndex.kt
package com.qeli.geo

import java.util.TreeMap

enum class GeoDomainType { Plain, Regex, RootDomain, Full }

data class GeoDomainRule(val type: GeoDomainType, val value: String)

/** Loaded subset of geosite.dat for selected tags. */
class GeoSiteIndex {
    private val byTag = TreeMap<String, MutableList<GeoDomainRule>>(String.CASE_INSENSITIVE_ORDER)

    fun matchAny(hostRaw: String, vararg tags: String): Boolean {
        if (hostRaw.isBlank()) return false
        val host = hostRaw.trim().trimEnd('.').lowercase()
        for (tag in tags) {
            val rules = byTag[tag] ?: continue
            if (rules.any { matchOne(host, it) }) return true
        }
        return false
    }

    /** True when [host] ends with `.<suffix>` or equals [suffix] (e.g. `.ru`, `.su`). */
    fun matchesDirectTld(hostRaw: String, suffixes: Collection<String>): Boolean {
        if (hostRaw.isBlank() || suffixes.isEmpty()) return false
        val host = hostRaw.trim().trimEnd('.').lowercase()
        for (suffix in suffixes) {
            val tld = suffix.trim().trim('.').lowercase()
            if (tld.isEmpty()) continue
            if (host == tld || host.endsWith(".$tld")) return true
        }
        return false
    }

    /**
     * Concrete domain names from loaded geosite rules (skips TLD-only wildcards like `ru`).
     * Used to pre-resolve A records into TUN exclude routes at connect time.
     */
    fun collectResolvableDomains(limit: Int): List<String> {
        require(limit > 0) { "limit must be positive" }
        val out = LinkedHashSet<String>()
        for (rules in byTag.values) {
            for (rule in rules) {
                val candidate = when (rule.type) {
                    GeoDomainType.Full -> rule.value
                    GeoDomainType.RootDomain -> rule.value
                    else -> continue
                }.trim().trimEnd('.').lowercase()
                if (candidate.isEmpty()) continue
                if (!candidate.contains('.')) continue
                out += candidate
                if (out.size >= limit) return out.toList()
            }
        }
        return out.toList()
    }

    private fun matchOne(host: String, rule: GeoDomainRule): Boolean {
        val v = rule.value.trim().lowercase()
        if (v.isEmpty()) return false
        return when (rule.type) {
            GeoDomainType.Full -> host == v
            GeoDomainType.RootDomain -> host == v || host.endsWith(".$v")
            GeoDomainType.Plain -> host.contains(v)
            GeoDomainType.Regex -> try { Regex(v).containsMatchIn(host) } catch (_: Exception) { false }
        }
    }

    companion object {
        fun matchesDirectTld(hostRaw: String, suffixes: Collection<String>): Boolean =
            GeoSiteIndex().matchesDirectTld(hostRaw, suffixes)

        fun load(path: java.io.File, tags: Collection<String>): GeoSiteIndex {
            val want = tags.map { it.lowercase() }.toHashSet()
            val bytes = path.readBytes()
            val index = GeoSiteIndex()
            val reader = ProtoReader(bytes)
            while (reader.hasMore) {
                val tag = reader.tryReadTag() ?: break
                val (field, wt) = tag
                if (field != 1 || wt != 2) { reader.skip(wt); continue }
                parseGeoSite(reader.readBytes(), want, index)
            }
            return index
        }

        private fun parseGeoSite(msg: ByteArray, want: Set<String>, index: GeoSiteIndex) {
            var code: String? = null
            val domains = ArrayList<GeoDomainRule>()
            val r = ProtoReader(msg)
            while (r.hasMore) {
                val tag = r.tryReadTag() ?: break
                val (field, wt) = tag
                when {
                    field == 1 && wt == 2 -> code = r.readString()
                    field == 2 && wt == 2 -> domains.add(parseDomain(r.readBytes()))
                    else -> r.skip(wt)
                }
            }
            val c = code ?: return
            if (c.lowercase() !in want) return
            index.byTag.getOrPut(c) { ArrayList() }.addAll(domains)
        }

        private fun parseDomain(msg: ByteArray): GeoDomainRule {
            var type = GeoDomainType.Plain
            var value = ""
            val r = ProtoReader(msg)
            while (r.hasMore) {
                val tag = r.tryReadTag() ?: break
                val (field, wt) = tag
                when {
                    field == 1 && wt == 0 -> type = when (r.readVarint().toInt()) {
                        1 -> GeoDomainType.Regex
                        2 -> GeoDomainType.RootDomain
                        3 -> GeoDomainType.Full
                        else -> GeoDomainType.Plain
                    }
                    field == 2 && wt == 2 -> value = r.readString()
                    else -> r.skip(wt)
                }
            }
            return GeoDomainRule(type, value)
        }
    }
}
